patches/api/0387-More-Teleport-API.patch

```diff
@@ -35,51 +35,100 @@ index 0000000000000000000000000000000000000000..7713a5b7e06da185685f18e79eae2c97
 new file mode 100644
-index 0000000000000000000000000000000000000000..0426ee8bd71142b6f933a479c0f2e5ef6043147d
+index 0000000000000000000000000000000000000000..61bf11978a712be805afb34506a4171ed1972777
 --- /dev/null
-+++ b/src/main/java/io/papermc/paper/entity/RelativeTeleportFlag.java
-@@ -0,0 +1,34 @@
++++ b/src/main/java/io/papermc/paper/entity/TeleportFlag.java
+@@ -0,0 +1,83 @@
 +package io.papermc.paper.entity;
 +
 +import org.bukkit.Location;
 +import org.bukkit.event.player.PlayerTeleportEvent;
++import org.jetbrains.annotations.ApiStatus;
 +
 +/**
-+ * Represents coordinates in a teleportation that should be handled relatively.
++ * Represents a flag that can be set on teleportation that may
++ * slightly modify the behavior.
 + *
-+ * @see org.bukkit.entity.Player#teleport(Location, PlayerTeleportEvent.TeleportCause, boolean, boolean, RelativeTeleportFlag...)
++ * @see EntityStateFlag
++ * @see RelativeTeleportFlag
 + */
-+@org.jetbrains.annotations.ApiStatus.Experimental
-+public enum RelativeTeleportFlag {
-+    /**
-+     * Represents the player's X coordinate
-+     */
-+    X,
-+    /**
-+     * Represents the player's Y coordinate
-+     */
-+    Y,
-+    /**
-+     * Represents the player's Z coordinate
-+     */
-+    Z,
++@ApiStatus.Experimental
++public sealed interface TeleportFlag permits TeleportFlag.EntityStateFlag, TeleportFlag.RelativeTeleportFlag {
++
 +    /**
-+     * Represents the player's yaw
++     * Note: These flags only work on {@link org.bukkit.entity.Player} entities.
++     * <p>
++     * Represents coordinates in a teleportation that should be handled relatively.
++     * <p>
++     * Coordinates of the location that the client should handle as relative teleportation
++     * Relative teleportation flags are only used client side, and cause the player to not lose velocity in that
++     * specific coordinate. The location of the teleportation will not change.
++     *
++     * @see org.bukkit.entity.Player#teleport(Location, PlayerTeleportEvent.TeleportCause, TeleportFlag...)
 +     */
-+    YAW,
++    @org.jetbrains.annotations.ApiStatus.Experimental
++    enum RelativeTeleportFlag implements TeleportFlag {
++        /**
++         * Represents the player's X coordinate
++         */
++        X,
++        /**
++         * Represents the player's Y coordinate
++         */
++        Y,
++        /**
++         * Represents the player's Z coordinate
++         */
++        Z,
++        /**
++         * Represents the player's yaw
++         */
++        YAW,
++        /**
++         * Represents the player's pitch
++         */
++        PITCH;
++    }
++
 +    /**
-+     * Represents the player's pitch
++     * Represents flags that effect the entity's state on
++     * teleportation.
 +     */
-+    PITCH;
++    @ApiStatus.Experimental
++    enum EntityStateFlag implements TeleportFlag {
```

Comment:
Adding this comment here because it involves the three ESF entries.
Maybe it's just me but I find the names do not communicate what effect they actually perform (or not); `IGNORE_OPEN_INVENTORY`, ignore what about it? Perhaps something like `RETAIN_...` is a bit more self-descriptive and, or some other wording; `IGNORE_PASSENGERS`, again, what about them? `RETAIN_...`/`KEEP_...` conveys more sense; same deal with `IGNORE_DISMOUNT`

Reply:
Yeah, it's a bit tricky I think. I do like the whole retain prefix.

"RETAIN_PASSENGERS", "RETAIN_OPEN_INVENTORY", "RETAIN_VEHICLE"

How do these sound?

Reply:
Yeah, I think that makes it more obvious what does/doesn't happen without recourse to the docs, (at least not to check the details about cross-world teleportation, that can't exactly be included in the name.)